From 123e194c7c2c43cf2da942a4f983d814deaf0f8d Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Tue, 29 Apr 2025 22:49:08 +0800
Subject: [PATCH 714/960] downstream: fs/resctrl: Add l2 mount option to enable
 L2 msc

The L2 msc function interface is hidden by default, and allow to enable
after adding mounting option of "-o l2" by user specified.

Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 fs/resctrl/ctrlmondata.c |  6 ++++++
 fs/resctrl/internal.h    |  1 +
 fs/resctrl/rdtgroup.c    | 16 ++++++++++++++++
 3 files changed, 23 insertions(+)

diff --git a/fs/resctrl/ctrlmondata.c b/fs/resctrl/ctrlmondata.c
index d2793e6407e8..025fbf91dfc6 100644
--- a/fs/resctrl/ctrlmondata.c
+++ b/fs/resctrl/ctrlmondata.c
@@ -311,6 +311,9 @@ static int rdtgroup_parse_resource(char *resname, char *tok,
 	struct resctrl_schema *s;
 
 	list_for_each_entry(s, &resctrl_schema_all, list) {
+		if (s->res->invisible)
+			continue;
+
 		if (!strcmp(resname, s->name) && rdtgrp->closid < s->num_closid)
 			return parse_line(tok, s, rdtgrp);
 	}
@@ -455,6 +458,9 @@ int rdtgroup_schemata_show(struct kernfs_open_file *of,
 		} else {
 			closid = rdtgrp->closid;
 			list_for_each_entry(schema, &resctrl_schema_all, list) {
+				if (schema->res->invisible)
+					continue;
+
 				if (closid < schema->num_closid)
 					show_doms(s, schema, closid);
 			}
diff --git a/fs/resctrl/internal.h b/fs/resctrl/internal.h
index 9e7ff00b8514..e1574ca71317 100644
--- a/fs/resctrl/internal.h
+++ b/fs/resctrl/internal.h
@@ -58,6 +58,7 @@ struct rdt_fs_context {
 	bool				enable_cdpl3;
 	bool				enable_mba_mbps;
 	bool				enable_debug;
+	bool				enable_l2;
 };
 
 static inline struct rdt_fs_context *rdt_fc2context(struct fs_context *fc)
diff --git a/fs/resctrl/rdtgroup.c b/fs/resctrl/rdtgroup.c
index b4fb125a7509..b2ed307beb58 100644
--- a/fs/resctrl/rdtgroup.c
+++ b/fs/resctrl/rdtgroup.c
@@ -2367,6 +2367,7 @@ static void rdt_disable_ctx(void)
 
 static int rdt_enable_ctx(struct rdt_fs_context *ctx)
 {
+	struct rdt_resource *r;
 	int ret = 0;
 
 	if (ctx->enable_cdpl2) {
@@ -2390,6 +2391,13 @@ static int rdt_enable_ctx(struct rdt_fs_context *ctx)
 	if (ctx->enable_debug)
 		resctrl_debug = true;
 
+	r = resctrl_arch_get_resource(RDT_RESOURCE_L2);
+	/* Only arm64 arch hides L2 resource by default */
+	if (IS_ENABLED(CONFIG_ARM64_MPAM) && !ctx->enable_l2)
+		r->invisible = true;
+	else
+		r->invisible = false;
+
 	return 0;
 
 out_cdpl3:
@@ -2612,6 +2620,7 @@ enum rdt_param {
 	Opt_cdpl2,
 	Opt_mba_mbps,
 	Opt_debug,
+	Opt_l2,
 	nr__rdt_params
 };
 
@@ -2620,6 +2629,7 @@ static const struct fs_parameter_spec rdt_fs_parameters[] = {
 	fsparam_flag("cdpl2",		Opt_cdpl2),
 	fsparam_flag("mba_MBps",	Opt_mba_mbps),
 	fsparam_flag("debug",		Opt_debug),
+	fsparam_flag("l2",		Opt_l2),
 	{}
 };
 
@@ -2648,6 +2658,9 @@ static int rdt_parse_param(struct fs_context *fc, struct fs_parameter *param)
 	case Opt_debug:
 		ctx->enable_debug = true;
 		return 0;
+	case Opt_l2:
+		ctx->enable_l2 = true;
+		return 0;
 	}
 
 	return -EINVAL;
@@ -2997,6 +3010,9 @@ static int mkdir_mondata_all(struct kernfs_node *parent_kn,
 		if (!r->mon_capable)
 			continue;
 
+		if (r->invisible)
+			continue;
+
 		ret = mkdir_mondata_subdir_alldom(kn, r, prgrp);
 		if (ret)
 			goto out_destroy;
-- 
2.43.0

